src/MergeContext.js

```diff
@@ -454,6 +454,214 @@ class BranchPosition
     }
 }
 
+// Forward iterator for fields in the 'name:value' format.
+class FieldsTokenizer
+{
+    constructor(str) {
+        this._lines = str.split('\n');
+    }
+
+    // Extracts the next line from the input and checks that it matches the
+    // 'attrName:attrValue' pattern.
+    // Returns the matched {attrName, attrValue} pair.
+    nextField() {
+        assert(!this.atEnd());
+        const line = this._lines.shift();
+
+        if (/^\s/.test(line))
+            throw new Error(`the line starts with a delimiter: '${line}'`);
+
+        assert(line.trim().length > 0);
+
+        const pos = line.search(':');
+        if (pos < 0)
+            throw new Error(`':' is missing in the line: '${line}'`);
+
+        return {name: line.substring(0, pos), value: line.substring(pos+1)};
+    }
+
+    // whether we are at the end of input
+    atEnd() {
+        // treat an empty line as the end of input
+        return this._lines.length === 0 || this._lines[0].trim().length === 0;
+    }
+
+    // returns the unparsed (yet) string
+    remaining() {
+        return this._lines.join('\n');
+    }
+}
+
+// computes future commit message from raw PR
+class CommitMessage
+{
+    constructor(rawPr, defaultAuthor) {
+        // the (required) commit message title
+        this._title = rawPr.title + ' (#' + rawPr.number + ')';
+        // the main description part, without header and trailer
+        this._body = undefined;
+        // The optional finalizing description part with GitHub-related attributes,
+        // separated from the body by an empty line.
+        this._trailer = undefined;
+        // The 'Authored-by' meta information extracted from the optional description header,
+        // separated from the body by an empty line.
+        this._author = undefined;
+        // Author in the {name, email, date} format.
+        // The default author will be used in the future commit if the 'Authored-by' attribute is missing.
+        this._defaultAuthor = defaultAuthor;
+
+        this._checkRaw(this._title);
+
+        if (rawPr.body !== undefined && rawPr.body !== null) {
+            const prDescription = rawPr.body.replace(/\r+\n/g, '\n');
+            this._checkRaw(prDescription);
+            this._parse(prDescription);
+        }
+    }
+
+    // complete message for the future commit
+    whole() {
+        assert(this._author !== null && this._body !== null && this._trailer !== null);
+        let message = this._title;
+        if (this._body)
```

Comment:
I thought that the code was written so that this._body was always set (because this._parseBody() was called for all successfully constructed objects). However, I now see that the constructor assumes that rawPr.body may be undefined or null (and does not call _parseBody() in those cases). I missed that. Sorry.

It feels like there is a problem somewhere:

* either the lack of PR description is not normal/expected and should result in CommitMessage constructor throwing
* or undefined/null PR description just implies an empty PR description, and the CommitMessage constructor should parse that empty description instead of not parsing at all (to avoid creating a third body state that we should worry about in whole() and possibly other methods).

I suspect we should restructure the code a little so that this._body is either null or a non-empty string (but never undefined). Alternatively, we can restructure so that this._body is always a string (never null/undefined) but can be an empty string. I think the first alternative is slightly better/clearer, but I am not sure. Please pick the best alternative and implement it.